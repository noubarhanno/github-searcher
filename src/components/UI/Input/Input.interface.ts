import { SpaceProps, WidthProps } from 'styled-system';
import { ChangeEvent } from 'react';
type Iinput = SpaceProps &
    WidthProps & {
        label: string;
        onChangeHandler: (e: ChangeEvent<HTMLInputElement>) => void;
        value: string;
    };

export default Iinput;
