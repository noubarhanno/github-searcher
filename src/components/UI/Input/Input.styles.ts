import styled from '@emotion/styled';
import { space, width } from 'styled-system';

export const InputBox = styled.div`
    display: flex;
    align-items: center;
    justify-content: space-between;
    height: 50px;
    width: 100%;
    color: ${(props: any) => props.theme.colors.text.light};
    padding: 10px 12px 9px;
    border-radius: ${(props: any) => props.theme.radii[2]}px;
    border: 1px solid ${(props: any) => props.theme.colors.text.light};
    background: white;
    &::after {
        background: white;
        position: absolute;
        top: 12px;
        left: 8px;
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        border-top: 10px solid #ffffff;
        width: 0;
        height: 0;
    }
`;

export const StyledInput = styled.input`
    width: 100%;
    border: none;
    outline: none;
    color: ${(props: any) => props.theme.colors.text.dark};
    font-weight: ${(props: any) => props.theme.fontWeights[4]};
    &::placeholder {
        font-weight: ${(props: any) => props.theme.fontWeights[4]};
        color: ${(props: any) => props.theme.colors.text.main};
    }
`;

export const InputWrpper = styled.div`
    position: relative;
    display: inline-block;
    ${space};
    ${width};
`;
