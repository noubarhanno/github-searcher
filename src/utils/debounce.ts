let timeout: NodeJS.Timeout | null;
export default (func: any, wait: number, immediate: boolean = false) => {
    return function (...args: any) {
        // @ts-ignore
        const context: any = this;
        const later = () => {
            timeout = null;
            if (!immediate) {
                func.apply(context, args);
            }
        };
        const callNow = immediate && !timeout;
        // @ts-ignore
        clearTimeout(timeout);
        // @ts-ignore
        timeout = setTimeout(later, wait);
        if (callNow) {
            func.apply(context, args);
        }
    };
};
