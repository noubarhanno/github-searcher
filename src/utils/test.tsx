import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import ThemeProvider from 'components/themeProvider';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';
import initialAppState from 'store/initialState';
import thunk from 'redux-thunk';

const AllTheProviders: React.FC = (props) => <ThemeProvider includeReset={false} {...props} />;

const customRender = (ui: React.ReactElement, options?: RenderOptions) => {
    return render(ui, { wrapper: AllTheProviders, ...options });
};

const renderWithRedux = (component: React.ReactNode, storeState = {}): any => {
    const getStore = createMockStore([thunk]);
    const store = getStore({ ...initialAppState, ...storeState });
    return {
        ...customRender(
            <Provider store={store}>
                <ThemeProvider includeReset={false}>{component}</ThemeProvider>
            </Provider>,
        ),
        store,
    };
};

/**
 * This method creates a mock store for testing purposes
 * @param state : any redux state
 */
export const getMockStore = (state?: any) => {
    const mockStore = createMockStore([thunk]);
    return mockStore({ ...initialAppState, ...state });
};

export * from '@testing-library/react';
export { customRender as render, renderWithRedux };
