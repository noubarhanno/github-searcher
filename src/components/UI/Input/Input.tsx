import React, { FunctionComponent, useRef } from 'react';
import { StyledInput, InputWrpper, InputBox } from './Input.styles';
import Iinput from './Input.interface';

const Input: FunctionComponent<Iinput> = ({ label, onChangeHandler, value, ...rest }) => {
    /**
     * Ref
     */
    const ref = useRef(null);

    return (
        <InputWrpper ref={ref} {...rest}>
            <InputBox>
                <StyledInput
                    data-testid="input-styled"
                    value={value}
                    placeholder={label}
                    onChange={onChangeHandler}
                    autoFocus
                />
            </InputBox>
        </InputWrpper>
    );
};

Input.defaultProps = {
    m: 1,
};

export default Input;
