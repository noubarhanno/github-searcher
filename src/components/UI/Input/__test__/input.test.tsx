import React from 'react';
import { render } from 'utils/test';
import Input from '../Input';

const spyFn: any = jest.fn();
describe('Input test suite', () => {
    test('Snapshot', () => {
        const { container } = render(<Input onChangeHandler={spyFn} label="test" value="test" />);
        expect(container).toMatchSnapshot();
    });

    test('should render the label as placeholder', () => {
        const { getByTestId } = render(<Input onChangeHandler={spyFn} label="test" value="test" />);
        const wrapper = getByTestId('input-styled');
        expect(wrapper).toHaveAttribute('placeholder', 'test');
    });
});
