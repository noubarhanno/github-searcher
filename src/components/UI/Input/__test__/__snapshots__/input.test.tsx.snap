// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input test suite Snapshot 1`] = `
<div>
  <div
    class="css-19yda7x"
  >
    <div
      class="css-fe2xgz"
    >
      <input
        class="css-1a8r2fa"
        data-testid="input-styled"
        placeholder="test"
        value="test"
      />
    </div>
  </div>
</div>
`;
